frontend/lib/decklist-parser.ts

```diff
@@ -0,0 +1,107 @@
+import lookup from "@/lib/api/spellbook-api";
+
+import type { FormattedApiResponse } from "@/lib/api/types";
+import type Card from "@/lib/api/models/card";
+
+type CombosInDecklist = {
+  combosInDecklist: FormattedApiResponse[];
+  potentialCombos: FormattedApiResponse[];
+  missingCardsForPotentialCombos: Card[];
+};
+
+type Deck = {
+  cards: string[];
+  numberOfCards: number;
+};
+
+export function convertDecklistToDeck(decklist: string): Deck {
+  let numberOfCardsInDeck = 0;
+
+  const cards = decklist
+    .split("\n")
+    .filter((cardName) => {
+      const normalizedCardName = cardName.trim();
+      // filter out empty values and values that start with //
+      // (common way to denote sections of a decklist, so it shouldn't
+      // be considered part of the decklist)
+      return normalizedCardName && !normalizedCardName.startsWith("//");
+    })
+    .map((cardName) => {
+      // this is a little complicated, but basically
+      // we pull off the quantity for the card (if any)
+      // when the card is in one of the following forms
+      // * 123 Card Name
+      // * 123x Card Name
+      // * 123X Card Name
+      // this is saved as the second match in the array
+      // returned by the match method, the card name is
+      // saved in the 3rd match
+      // it's technically possible for match to return null
+      // so we need to default to an empty array to destructure
+      // the variables
+      const [, quantityAsString, name] =
+        cardName.match(/^(\d*[xX]?\s)?(.*)$/) || [];
+      // if the quantity doesn't exist, then we default to 1
+      // and we remove the x (lowercase or uppercase)
+      // before casting it to a number
+      const quantity = Number((quantityAsString || "1").replace(/x/i, ""));
+
+      numberOfCardsInDeck += quantity;
+
+      return name;
+    })
+    .map((cardName) => {
+      // removes set name or collector data, which we assume
+      // is anything after a space followed by an open parenthesis
+      // NOTE: there are a handful of magic cards that use ( in the
+      // the name, but they are all silver border so far and unlikely
+      // to be present in the Commander Spellbook database
+      return cardName.split(" (")[0].trim();
+    });
+
+  return {
+    numberOfCards: numberOfCardsInDeck,
+    cards,
+  };
+}
```

Comment:
I typed this in a textarea so there may be small errors/typos in the code, but this is the simplification I would consider. The regular expression should really be moved out of the loop and created as a constant in the header of this file.

```suggestion
export const convertDecklistToDeck = (decklist: string): Deck => decklist
  .split('\n')
  .reduce(
    (result, line) => {
      /*
       * Accepts lines in the following formats and all variations in between:
       * 123 Card Name
       * 123x Card Name
       * Card Name (set name)
       */
      const match = line.match(/^\s*(?:(?<count>\d+)x?\s+)?(?<name>[^\(]+)/i);
      
      if (match) {
        result.numberOfCards += Number(match.groups.count || 1);
        result.cards.push(match.groups.name);
      }
      
      return result;
    },
    {
      cards: [],
      numberOfCards: 0,
    },
  )
;
```

Reply:
I always forget the syntax for named matches in regexes. Thanks for the reminder.

Reply:
I tried fiddling with your regex to make it work, but couldn't. If you want to clone it down and figure it out, happy to take another look to see if it'll improve the situation. 

As a refactor, I pulled out most of the logic into smaller named functions.

Reply:
> I tried fiddling with your regex to make it work, but couldn't.

Are there tests for this? I didn't see any.

What's not working exactly?

Reply:
Yeah, there are tests for this located at `frontend/test/lib/decklist-parser.test.ts`

Various tests just wouldn't pass with the regex as is involving correct counts and parsing out card names with the set code stuff. Regex is a weak point for me, so I tried modifying it a bit to make it work, and I just got frustrated with it and went back to this.

If it can be done all in one go with a well-documented Regex, cool, but I'm not bending over backwards to figure out the exact incantation to do so.

Reply:
Finally coming back to this. I created #330 forked from this branch